### PR TITLE
Added alternative payments CS1.

### PR DESCRIPTION
**Pull request by Armando-Ibarra**
Added support for alternative payments source for `Ideal`, `Klarna`, `Oxxo`, `PagoFacil`, `RapiPago`, `SEPA`, which includes the following clients: `IdealClient`, `KlarnaClient`,` OxxoClient`, `PagoFacilClient`, `RapiPagoClient`, `SepaClient`.

Added a refactor to distinguish source of payment in PaymentSource

### CODE REVIEW

File: src/main/java/com/checkout/apm/ideal/IdealCountry.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.checkout.apm.ideal;
+
+import lombok.Builder;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+@Builder
+public class IdealCountry {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/ideal/IdealCountry.java:10]: Class should be `final`. If it is a response class, please remove `@Builder`.

File: src/main/java/com/checkout/apm/ideal/Issuer.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.checkout.apm.ideal;
+
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public final class Issuer {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/ideal/Issuer.java:8]: If it is a response class, please remove `@Builder`.

File: src/main/java/com/checkout/apm/klarna/KlarnaClientImpl.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    private static final String ORDERS = "orders";
+    private static final String CAPTURES = "captures";
+    private static final String VOIDS = "voids";
+
+    public KlarnaClientImpl(final ApiClient apiClient, final CheckoutConfiguration configuration) {
+        super(apiClient, PublicKeyCredentials.fromConfiguration(configuration));
+    }
+
+    @Override
+    public CompletableFuture<CreditSessionResponse> createSession(final CreditSessionRequest creditSessionRequest) {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/klarna/KlarnaClientImpl.java:27]: Should this operation be called `createCreditSession`? Just to be similar to `getCreditSession`.

File: src/main/java/com/checkout/apm/klarna/OrderCaptureRequest.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.List;
+import java.util.Map;
+
+@Data
+@Builder
+public final class OrderCaptureRequest {
+
+    private final PaymentSource type = PaymentSource.KLARNA;
+
+    private Integer amount;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/klarna/OrderCaptureRequest.java:19]: Better changing `amount` type to `Long`.

File: src/main/java/com/checkout/apm/klarna/OrderCaptureRequest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+        private String description;
+
+        private List<KlarnaProduct> products;
+
+        @SerializedName("shipping_info")
+        private List<ShippingInfo> shippingInfo;
+
+        @SerializedName("shipping_delay")
+        private Integer shipping_delay;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/klarna/OrderCaptureRequest.java:41]: Please change `shipping_delay` to `shippingDelay`, and type to `Long`.

File: src/main/java/com/checkout/apm/klarna/PaymentMethod.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import com.google.gson.annotations.SerializedName;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+public final class PaymentMethod {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/klarna/PaymentMethod.java:12]: If this is a request class, please add `@Builder` and `@AllArgsConstructor`.

File: src/main/java/com/checkout/apm/klarna/PaymentMethod.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    private String identifier;
+
+    private String name;
+
+    @SerializedName("asset_urls")
+    private List<AssetUrl> assetUrls;
+
+    @Data
+    @Builder
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/klarna/PaymentMethod.java:22]: If this is a response class, please remove unnecessary annotations.

File: src/main/java/com/checkout/apm/oxxo/OxxoClientImpl.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+public class OxxoClientImpl extends AbstractClient implements OxxoClient {
+
+    private static final String OXXO_PATH = "/apms/oxxo/payments";
+
+    public OxxoClientImpl(final ApiClient apiClient, final CheckoutConfiguration configuration) {
+        super(apiClient, SecretKeyCredentials.fromConfiguration(configuration));
+    }
+
+    @Override
+    public CompletableFuture<Void> succeed(final String paymentId) {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/oxxo/OxxoClientImpl.java:19]: Please add validation for `paymentId`.

File: src/main/java/com/checkout/apm/oxxo/OxxoClientImpl.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        return apiClient.postAsync(buildPath(OXXO_PATH, paymentId, "succeed"), apiCredentials, Void.class, null, null);
+    }
+
+    @Override
+    public CompletableFuture<Void> expire(final String paymentId) {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/oxxo/OxxoClientImpl.java:24]: Same :)

File: src/main/java/com/checkout/apm/oxxo/OxxoClient.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.checkout.apm.oxxo;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface OxxoClient {
+
+    CompletableFuture<Void> succeed(String id);
+
+    CompletableFuture<Void> expire(String id);
+
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/oxxo/OxxoClient.java:10]: Please change both parameters to `paymentId` to match the implementation class (and be more informative).

File: src/main/java/com/checkout/apm/pagofacil/PagoFacilClient.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.checkout.apm.pagofacil;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface PagoFacilClient {
+
+    CompletableFuture<Void> succeed(String id);
+
+    CompletableFuture<Void> expire(String id);
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/pagofacil/PagoFacilClient.java:9]: Same. Please change parameters to `paymentId`.

File: src/main/java/com/checkout/apm/pagofacil/PagoFacilImpl.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    private static final String PAGO_FACIL_PATH = "/apms/pagofacil/payments";
+
+    public PagoFacilImpl(final ApiClient apiClient, final CheckoutConfiguration configuration) {
+        super(apiClient, SecretKeyCredentials.fromConfiguration(configuration));
+    }
+
+    @Override
+    public CompletableFuture<Void> succeed(final String paymentId) {
+        validateParams("paymentId",paymentId);
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/pagofacil/PagoFacilImpl.java:22]: Please format the code :)

File: src/main/java/com/checkout/apm/rapipago/RapiPagoClient.java [hunk 13]
```diff
@@ -0,0 +1,11 @@
+package com.checkout.apm.rapipago;
+
+import java.util.concurrent.CompletableFuture;
+
+public interface RapiPagoClient {
+
+    CompletableFuture<Void> succeed(String id);
+
+    CompletableFuture<Void> expire(String id);
+
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/rapipago/RapiPagoClient.java:10]: Same :)

File: src/main/java/com/checkout/apm/sepa/MandateResponse.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    private String customerId;
+
+    @SerializedName("first_name")
+    private String firstName;
+
+    @SerializedName("last_name")
+    private String lastName;
+
+    @SerializedName("address_line1")
+    private String addressLine1;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/sepa/MandateResponse.java:24]: Only `addressLine1` is available in this response? Usually 2 addresses are returned.

File: src/main/java/com/checkout/common/PaymentSource.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
+package com.checkout.common;
+
+import com.google.gson.annotations.SerializedName;
+
+public enum PaymentSource {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/common/PaymentSource.java:5]: There's already an enum `PaymentSourceType` in `common`. Should we reuse it?

File: src/main/java/com/checkout/common/PaymentSource.java [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    IDEAL,
+    @SerializedName("oxxo")
+    OXXO,
+    @SerializedName("pagofacil")
+    PAGO_FACIL,
+    @SerializedName("rapipago")
+    RAPI_PAGO,
+    @SerializedName("klarna")
+    KLARNA,
+    @SerializedName("id")
```
Comment (rsaestrela) [inline src/main/java/com/checkout/common/PaymentSource.java:35]: Please confirm if `id` is the correct code for `SEPA`.

File: src/main/java/com/checkout/payments/CardSource.java [hunk 17]
```diff
@@ -14,7 +15,7 @@
 @AllArgsConstructor
 public class CardSource implements RequestSource {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/CardSource.java:17]: If `@Override public String getType()` is never used, I believe we can remove it from the interface. We can still use the interface (marker interface) for polymorphism if strictly necessary, or just deprecate it to be removed in a future version.

File: src/main/java/com/checkout/payments/apm/BalotoSource.java [hunk 18]
```diff
@@ -23,18 +22,9 @@
...
 
-    @Data
-    @Builder
-    @NoArgsConstructor
-    @AllArgsConstructor
-    public static class Payer {
-        private String name;
-        private String email;
-    }
-
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/apm/BalotoSource.java:34]: I think `Payer` should still be here. I like the idea or reusing it, but `BalotoSource` payer does not have the attribute `document` for example. In practice it should be ok, but I would prefer to have clear/specific payer objects.

File: src/main/java/com/checkout/payments/apm/IdealSource.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
+package com.checkout.payments.apm;
+
+import com.checkout.common.PaymentSource;
+import com.checkout.payments.RequestSource;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+public final class IdealSource implements RequestSource {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/apm/IdealSource.java:10]: Please add constructor annotations.

File: src/main/java/com/checkout/payments/apm/KlarnaSource.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    @SerializedName("authorization_token")
+    private final String authorizationToken;
+
+    private final String locale;
+
+    @SerializedName("purchase_country")
+    private final CountryCode purchaseCountry;
+
+    @SerializedName("tax_amount")
+    private final Integer taxAmount;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/apm/KlarnaSource.java:30]: Please change to Long and make the attributes non-final (just for consistency). Please do the same on all `RequestSource` classes (even the ones already implemented).

Immutability is technically important, but making these attributes non-final will probably be a more flexible approach for the users of the SDK.

File: src/main/java/com/checkout/payments/apm/KlarnaSource.java [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+
+        @SerializedName("street_address2")
+        private String streetAddress2;
+
+        @SerializedName("postal_code")
+        private String postalCode;
+
+        private String city;
+
+        private String phone;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/apm/KlarnaSource.java:61]: Please double-check is this is a String (usually is the class `Phone`).

File: src/main/java/com/checkout/payments/apm/OxxoSource.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+@Data
+@Builder
+public final class OxxoSource implements RequestSource {
+
+    private final PaymentSource type = PaymentSource.OXXO;
+
+    @SerializedName("integration_type")
+    private final IntegrationType integrationType = IntegrationType.REDIRECT;
+
+    private final CountryCode country;
```
Comment (rsaestrela) [inline src/main/java/com/checkout/payments/apm/OxxoSource.java:19]: In general please make request classes attributes non-final. The `type` can be final, of course.

**Review by rsaestrela — CHANGES_REQUESTED** [review]
Don't forget the documentation! :) Nice work!!!

File: src/main/java/com/checkout/apm/sepa/MandateResponse.java (hunk 14 above)
Comment (Armando-Ibarra) [reply src/main/java/com/checkout/apm/sepa/MandateResponse.java:24]: Yes, it only expects one https://docs.checkout.com/payments/payment-methods/direct-debit/sepa-direct-debit#SEPADirectDebit-CreateanewSDDpaymentsourcecreatepaymentsource

File: src/main/java/com/checkout/common/PaymentSource.java (hunk 16 above)
Comment (Armando-Ibarra) [reply src/main/java/com/checkout/common/PaymentSource.java:35]: Yes it is

File: src/main/java/com/checkout/payments/apm/KlarnaSource.java (hunk 21 above)
Comment (Armando-Ibarra) [reply src/main/java/com/checkout/payments/apm/KlarnaSource.java:61]: Yes it is only a String https://docs.checkout.com/payments/payment-methods/invoice-and-pay-later/klarna#Klarna-Placetheorderplaceorder

**Comment by Armando-Ibarra** [comment]
@rsaestrela I was not able to answers all the comments but thank you for the feedback.

**Review by rsaestrela — CHANGES_REQUESTED** [review]

File: src/main/java/com/checkout/apm/sepa/MandateResponse.java [hunk 23]
```diff
@@ -0,0 +1,50 @@
+package com.checkout.apm.sepa;
+
+import com.checkout.common.CountryCode;
+import com.checkout.common.Resource;
+import com.google.gson.annotations.SerializedName;
+import lombok.Data;
+
+@Data
+public final class MandateResponse extends Resource {
```
Comment (rsaestrela) [inline src/main/java/com/checkout/apm/sepa/MandateResponse.java:9]: Please add missing annotations (EqualsAndHashCode, etc), to this class.

**Pull request merged**